### PR TITLE
feat(SM): MCOL-5785 Add timeout options for S3Storage

### PR DESCRIPTION
**Pull request by mariadb-AlexeyAntipovsky**

### CODE REVIEW

File: storage-manager/src/S3Storage.cpp [hunk 1]
```diff
@@ -99,7 +102,7 @@ const char* s3err_msgs[] = {"All is well",
                             "Authentication failed, token has expired",
                             "Configured bucket does not match endpoint location"};
 
-S3Storage::ScopedConnection::ScopedConnection(S3Storage* s, ms3_st* m) : s3(s), conn(m)
+S3Storage::ScopedConnection::ScopedConnection(S3Storage* s, shared_ptr<S3Storage::Connection> m) : s3(s), conn(m)
```
Comment (drrtuy) [inline storage-manager/src/S3Storage.cpp:105]: Why boost and not std::shared_ptr @mariadb-AlexeyAntipovsky ?

File: storage-manager/src/S3Storage.cpp [hunk 2]
```diff
@@ -210,19 +215,27 @@ S3Storage::S3Storage(bool skipRetry) : skipRetryableErrors(skipRetry)
...
+  if (!timeout.empty())
+  {
+      operationTimeout = stof(timeout);
+  }
 
   ms3_library_init();
   if (libs3_debug == "enabled")
   {
-    ms3_debug();
+    ms3_debug(1);
```
Comment (drrtuy) [inline storage-manager/src/S3Storage.cpp:230]: Really? It didn't work previously?

File: storage-manager/src/S3Storage.cpp [hunk 3]
```diff
@@ -361,29 +374,29 @@ int S3Storage::getObject(const string& _sourceKey, std::shared_ptr<uint8_t[]>* d
 {
   uint8_t err;
   size_t len = 0;
-  uint8_t* _data = NULL;
+  uint8_t* _data = nullptr;
   string sourceKey = prefix + _sourceKey;
 
-  ms3_st* creds = getConnection();
+  auto creds = getConnection();
```
Comment (drrtuy) [inline storage-manager/src/S3Storage.cpp]: The variable `creds` is a result of `getConnection`. I am either stupid or the variable name is misguiding.

File: storage-manager/src/TerminateIOTask.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+#include "TerminateIOTask.h"
+#include "SMLogging.h"
+#include "messageFormat.h"
+#include "src/CloudStorage.h"
+#include <errno.h>
+#include <string.h>
+#include <iostream>
+
+using namespace std;
```
Comment (drrtuy) [inline storage-manager/src/TerminateIOTask.cpp]: Can we get away w/o this using?

File: storage-manager/src/S3Storage.cpp (hunk 2 above)
Comment (mariadb-AlexeyAntipovsky) [reply storage-manager/src/S3Storage.cpp:230]: Previously libmarias3 submodule pointed to a commit before that API change

File: storage-manager/src/S3Storage.cpp (hunk 3 above)
Comment (mariadb-AlexeyAntipovsky) [reply storage-manager/src/S3Storage.cpp]: Good point, thanks.

File: storage-manager/include/messageFormat.h [hunk 5]
```diff
@@ -300,6 +302,46 @@ struct copy_cmd
...
+/*
+    LIST_IOTASKS
+    ------------
+    command format:
+    1-byte opcode
+
+    response format:
+    4-byte num elements|(8-byte id|double runningTime) * num elements
+*/
+struct list_iotask_cmd
```
Comment (drrtuy) [inline storage-manager/include/messageFormat.h:314]: Should be camel case.

File: storage-manager/src/ListIOTask.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
+/* Copyright (C) 2019 MariaDB Corporation
```
Comment (drrtuy) [inline storage-manager/src/ListIOTask.cpp]: Should be 2024.

**Review by drrtuy — COMMENTED** [review]
Untested ack except some cosmetic changes

File: storage-manager/src/S3Storage.cpp (hunk 1 above)
Comment (mariadb-AlexeyAntipovsky) [reply storage-manager/src/S3Storage.cpp:105]: It is `std::shared_ptr`

**Comment by mariadb-LeonidFedorov** [comment]
I have added the commit, that should fix the build of debian based packages, PTAL

**Review by drrtuy — APPROVED** [review]
untestedACK

**Pull request merged**